### PR TITLE
feat(style): hide dashboard header by url parameter

### PR DESCRIPTION
**Pull request by simcha90**
### SUMMARY
- Hide dashboard header with name and edit button dependent on url parameter `standalone `
e.g. http://localhost:8088/superset/dashboard/11/?standalone=true
Here supported options (according proposal of @ktmud ):
a. `standalone=true` (for backward compatibility) or  `standalone=1` hide only navbar
b. `standalone=2` hide navbar and title
c. `standalone=false` (for backward compatibility) or  `no param` show all
- Add global constants for url params

Reason for PR: When dashboard is embedded as standalone there is also a use-case to hide the dashboard header

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

https://user-images.githubusercontent.com/56388545/107141317-a1460780-6930-11eb-9c7a-7aee26d3d777.mov

### TEST PLAN
- Test according 1. flows 
- Test filter bar of native filters working as expected

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by junlincc** [comment]
thanks for the PR @simcha90 ! User will still be able to add filters after hiding header and edit button. is this intended? 
@mihir174 I think this is a valid user need - sharing a dashboard to other that should not be editing it. this solution to me is a bit "hacky" though. any idea?

**Comment by mihir174** [comment]
Agreed that this is a bit hacky but it's a good first step. We need to think about security/view permission configurations on embed links. 
I like the way Google, for example, deals with link sharing:
Screen Shot 2021-02-03 at 4 31 16 PM

**Comment by ktmud** [comment]
Why not call the parameter `standalone`, too, just to be consistent with charts?

**Comment by simcha90** [comment]
> Why not call the parameter `standalone`, too, just to be consistent with charts?

`standalone` - hides only top menu bar
`hide_dashboard_hide` - hides only header with edit button

I think it dependent on user needs we need give both options to manage dashboard view

**Comment by ktmud** [comment]
Why would you want to keep Superset navs while hiding dashboard title? Wouldn't users be able to navigate to other Superset pages in the embedded dashboard? I'm finding it hard to imagine a scenario where this would be desired.

**Comment by amitmiran137** [comment]
@ktmud I agree that the use-case of standalone=false and hide_dashboard_hide=true does not exist
but there are 2 cases that do exist
1.  standalone=true & hide_dashboard_hide=false
2. standalone =true & hide_dashboard_hide=true

I can agree that by just setting hide_dashboard_hide=true we can enable standalone but not sure we must enforce it
WDYT?

**Comment by ktmud** [comment]
How about we introduce two "standalone modes"?

- `standelone=true` or `standlone=1`: hide navs only
- `standalone=2`: hide navs and title

Let's not give users the choice to hide just the dashboard title. If the request comes back up with a reasonable business need to support it, then we can add it as `standlone=3`.

**Comment by simcha90** [comment]
> How about we introduce two "standalone modes"?
> 
> * `standelone=true` or `standlone=1`: hide navs only
> * `standalone=2`: hide navs and title
> 
> Let's not give users the choice to hide just the dashboard title. If the request comes back up with a reasonable business need to support it, then we can add it as `standlone=3`.

done

File: superset/utils/core.py [hunk 1]
```diff
@@ -252,6 +252,12 @@ class ReservedUrlParameters(str, Enum):
     STANDALONE = "standalone"
     EDIT_MODE = "edit"
 
+    @classmethod
+    def is_standalone_mode(cls) -> Any:
+        standalone_param = request.args.get(ReservedUrlParameters.STANDALONE.value)
+        standalone = standalone_param != "false" and standalone_param is not None
```
Comment (ktmud) [inline superset/utils/core.py:258]:
```suggestion
        standalone = standalone_param and standalone_param != "false" and standalone_param != "0"
```

File: superset-frontend/src/dashboard/components/DashboardBuilder.jsx [hunk 2]
```diff
@@ -225,58 +227,72 @@ class DashboardBuilder extends React.Component {
 
     const childIds = topLevelTabs ? topLevelTabs.children : [DASHBOARD_GRID_ID];
 
-    const barTopOffset = HEADER_HEIGHT + (topLevelTabs ? TABS_HEIGHT : 0);
+    const hideDashboardHeader =
+      getUrlParam(URL_PARAMS.standalone, 'number') === 2;
```
Comment (ktmud) [inline superset-frontend/src/dashboard/components/DashboardBuilder.jsx:231]: Maybe make `1` and `2` a constant/enum somewhere?

```ts
enum DashboardStandaloneMode {
  hideNav = 1,
  hideNavAndTitle = 2,
}
```

**Review by ktmud — COMMENTED** [review]
Some code style suggestions, most are non-blocking.

File: superset-frontend/src/dashboard/util/getDashboardUrl.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  const hashSection = hash ? `#${hash}` : '';
+
+  return `${pathname}?${newSearchParams.toString()}${hashSection}`;
+}
+
+export type UrlParamType = 'string' | 'number' | 'boolean';
+export function getUrlParam(paramName: string, type: 'string'): string;
+export function getUrlParam(paramName: string, type: 'number'): number;
+export function getUrlParam(paramName: string, type: 'boolean'): boolean;
+export function getUrlParam(paramName: string, type: UrlParamType): unknown {
```
[the remark this answers is not in the document]
Comment (simcha90) [reply superset-frontend/src/dashboard/util/getDashboardUrl.ts:50]: moved it here: `superset-frontend/src/utils/common.ts`

File: superset-frontend/src/dashboard/util/getDashboardUrl.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  const urlParam = new URLSearchParams(window.location.search.substring(1)).get(
```
[the remark this answers is not in the document]
Comment (simcha90) [reply superset-frontend/src/dashboard/util/getDashboardUrl.ts:51]: done

File: superset-frontend/src/dashboard/util/getDashboardUrl.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        return null;
+      }
+      if (urlParam === 'true') {
+        return 1;
+      }
+      if (urlParam === 'false') {
+        return 0;
+      }
+      // eslint-disable-next-line no-case-declarations
+      const parsedNumber = parseInt(urlParam, 10);
```
[the remark this answers is not in the document]
Comment (simcha90) [reply superset-frontend/src/dashboard/util/getDashboardUrl.ts:66]: done

File: superset/utils/core.py [hunk 6]
```diff
@@ -252,6 +252,12 @@ class ReservedUrlParameters(str, Enum):
     STANDALONE = "standalone"
     EDIT_MODE = "edit"
 
+    @classmethod
+    def is_standalone_mode(cls) -> Any:
```
[the remark this answers is not in the document]
Comment (simcha90) [reply superset/utils/core.py:256]: done

File: superset/views/core.py [hunk 7]
```diff
@@ -400,9 +401,11 @@ def slice(self, slice_id: int) -> FlaskResponse:  # pylint: disable=no-self-use
...
             parse.quote(json.dumps({"slice_id": slice_id}))
         )
-        param = utils.ReservedUrlParameters.STANDALONE.value
-        if request.args.get(param) == "true":
-            endpoint += f"&{param}=true"
+
+        is_standalone_mode = ReservedUrlParameters.is_standalone_mode()
+        if is_standalone_mode:
+            param = utils.ReservedUrlParameters.STANDALONE.value
+            endpoint += f"&{param}={is_standalone_mode}"
```
[the remark this answers is not in the document]
Comment (simcha90) [reply superset/views/core.py:408]: done

File: superset/utils/core.py (hunk 1 above)
Comment (simcha90) [reply superset/utils/core.py:258]: done

File: superset-frontend/src/dashboard/components/DashboardBuilder.jsx (hunk 2 above)
Comment (simcha90) [reply superset-frontend/src/dashboard/components/DashboardBuilder.jsx:231]: done

**Review by ktmud — COMMENTED** [review]
Code LGTM, just one comment about typing.

I'm also seeing this gap for native filters. Is this expected?

image

File: superset-frontend/src/utils/common.ts [hunk 8]
```diff
@@ -44,7 +44,13 @@ export function getParamFromQuery(query, param) {
...
 }
 
-export function storeQuery(query) {
+export function storeQuery(query: {
+  dbId: any;
+  title: any;
+  schema: any;
+  autorun: any;
+  sql: any;
+}) {
```
[the remark this answers is not in the document]
Comment (simcha90) [reply superset-frontend/src/utils/common.ts:53]: 👍 did as second proposal, created file for `urlUtils`

**Comment by simcha90** [comment]
image
image

@ktmud  I can't reproduce UI bug that you sent, also Chrome also Firefox

**Comment by simcha90** [comment]
@ktmud 👍 good catch about UI bug, fixed

File: superset-frontend/src/explore/exploreUtils.js [hunk 9]
```diff
@@ -99,8 +100,8 @@ export function getExploreLongUrl(
     search[key] = extraSearch[key];
   });
   search.form_data = safeStringify(formData);
-  if (endpointType === 'standalone') {
-    search.standalone = 'true';
+  if (endpointType === URL_PARAMS.standalone) {
+    search.standalone = '1';
```
Comment (ktmud) [inline superset-frontend/src/explore/exploreUtils.js:104]: Could you use the constant `DashboardStandaloneMode` here as well?

Ditto in the spec files.

**Review by ktmud — APPROVED** [review]
LGTM with one last suggestion.

Thanks for all the iterations!

**Comment by ktmud** [comment]
cc @junlincc if you want to do a final round of QA & product sign-off.

File: superset-frontend/src/explore/exploreUtils.js (hunk 9 above)
Comment (simcha90) [reply superset-frontend/src/explore/exploreUtils.js]: fixed

File: superset-frontend/spec/javascripts/explore/utils_spec.jsx [hunk 10]
```diff
@@ -32,6 +32,7 @@ import {
 } from 'src/explore/dateFilterUtils';
 import * as hostNamesConfig from 'src/utils/hostNamesConfig';
 import { getChartMetadataRegistry } from '@superset-ui/core';
+import { DashboardStandaloneMode } from '../../../src/dashboard/util/constants';
```
Comment (villebro) [inline superset-frontend/spec/javascripts/explore/utils_spec.jsx:35]:
```suggestion
import { DashboardStandaloneMode } from 'src/dashboard/util/constants';
```

File: superset-frontend/spec/javascripts/dashboard/util/getDashboardUrl_spec.js [hunk 11]
```diff
@@ -18,28 +18,51 @@
...
+      '',
+      DashboardStandaloneMode.HIDE_NAV,
+    );
     expect(urlWithStandalone).toBe(
-      'path?preselect_filters=%7B%2235%22%3A%7B%22key%22%3A%5B%22value%22%5D%7D%7D&standalone=true',
+      `path?preselect_filters=%7B%2235%22%3A%7B%22key%22%3A%5B%22value%22%5D%7D%7D&standalone=${DashboardStandaloneMode.HIDE_NAV}`,
+    );
+  });
+
+  it('should encode filters with missed standalone', () => {
```
Comment (villebro) [inline superset-frontend/spec/javascripts/dashboard/util/getDashboardUrl_spec.js:62]:
```suggestion
  it('should encode filters with missing standalone', () => {
```

File: superset-frontend/spec/javascripts/explore/components/EmbedCodeButton_spec.jsx [hunk 12]
```diff
@@ -26,7 +26,8 @@ import configureStore from 'redux-mock-store';
 import fetchMock from 'fetch-mock';
 import EmbedCodeButton from 'src/explore/components/EmbedCodeButton';
 import * as exploreUtils from 'src/explore/exploreUtils';
-import * as common from 'src/utils/common';
+import * as urlUtils from 'src/utils/urlUtils';
+import { DashboardStandaloneMode } from '../../../../src/dashboard/util/constants';
```
Comment (villebro) [inline superset-frontend/spec/javascripts/explore/components/EmbedCodeButton_spec.jsx:30]:
```suggestion
import { DashboardStandaloneMode } from 'src/dashboard/util/constants';
```

**Review by villebro — APPROVED** [review]
A few non blocking nits

**Pull request merged**